File: star_map.go
// star_map.go
package main

import (
	"flag"
	"fmt"
	"math"
	"os"
	"strings"
)

type Star struct {
	X    int     `json:"x"`
	Y    int     `json:"y"`
	Name string  `json:"name"`
	Mag  float64 `json:"mag"`
}

type Constellation struct {
	Name          string   `json:"name"`
	Abbr          string   `json:"abbr"`
	Hemisphere    string   `json:"hemisphere"`
	Season        string   `json:"season"`
	BrightestStar string   `json:"brightest_star"`
	Mythology     string   `json:"mythology"`
	NotableObjs   string   `json:"notable_objects"`
	Stars         []Star   `json:"stars"`
	Connections   [][2]int `json:"connections"`
}

var constellations = []Constellation{
	{
		Name:          "Orion",
		Abbr:          "Ori",
		Hemisphere:    "both",
		Season:        "winter",
		BrightestStar: "Betelgeuse (Alpha Orionis, 0.42 mag)",
		Mythology:     "Orion, a mighty hunter in Greek mythology, was placed among the stars by Zeus. He is depicted as a hunter with his belt, sword, and club.",
		NotableObjs:   "Orion Nebula (M42), Horsehead Nebula, Barnard's Loop",
		Stars: []Star{
			{4, 1, "Betelgeuse", 0.42},
			{5, 4, "Bellatrix", 1.64},
			{3, 5, "Mintaka", 2.23},
			{5, 5, "Alnilam", 1.69},
			{7, 5, "Alnitak", 1.88},
			{2, 8, "Saiph", 2.06},
			{6, 9, "Rigel", 0.12},
		},
		Connections: [][2]int{{0, 1}, {0, 2}, {0, 3}, {0, 4}, {0, 5}, {0, 6}, {1, 2}, {2, 3}, {3, 4}, {4, 5}, {5, 6}, {2, 4}},
	},
	{
		Name:          "Ursa Major",
		Abbr:          "UMa",
		Hemisphere:    "north",
		Season:        "spring",
		BrightestStar: "Alioth (Epsilon Ursae Majoris, 1.77 mag)",
		Mythology:     "Ursa Major represents the Great Bear. In Greek mythology, Callisto was transformed into a bear by Hera and placed in the sky by Zeus.",
		NotableObjs:   "M81, M82 (Bode's Galaxy and Cigar Galaxy), Owl Nebula (M97)",
		Stars: []Star{
			{1, 1, "Dubhe", 1.79},
			{3, 2, "Merak", 2.34},
			{2, 4, "Phecda", 2.41},
			{4, 4, "Megrez", 3.31},
			{5, 5, "Alioth", 1.77},
			{7, 4, "Mizar", 2.04},
			{9, 3, "Alkaid", 1.86},
		},
		Connections: [][2]int{{0, 1}, {1, 2}, {2, 3}, {3, 4}, {4, 5}, {5, 6}},
	},
	{
		Name:          "Cassiopeia",
		Abbr:          "Cas",
		Hemisphere:    "north",
		Season:        "autumn",
		BrightestStar: "Schedar (Alpha Cassiopeiae, 2.24 mag)",
		Mythology:     "Cassiopeia was the vain queen of Ethiopia who boasted about her beauty, leading to the sacrifice of her daughter Andromeda. She was placed in the sky as punishment.",
		NotableObjs:   "Cassiopeia A (supernova remnant), NGC 457 (open cluster)",
		Stars: []Star{
			{2, 1, "Segin", 3.37},
			{3, 3, "Ruchbah", 2.68},
			{5, 2, "Schedar", 2.24},
			{6, 4, "Navi", 2.47},
			{8, 3, "Caph", 2.28},
		},
		Connections: [][2]int{{0, 1}, {1, 2}, {2, 3}, {3, 4}, {0, 2}, {2, 4}},
	},
	{
		Name:          "Scorpius",
		Abbr:          "Sco",
		Hemisphere:    "south",
		Season:        "summer",
		BrightestStar: "Antares (Alpha Scorpii, 0.96 mag)",
		Mythology:     "Scorpius represents the scorpion that killed Orion. According to mythology, the scorpion was placed on the opposite side of the sky from Orion.",
		NotableObjs:   "Antares (red supergiant), Ptolemy's Cluster (M7), Butterfly Cluster (M6)",
		Stars: []Star{
			{2, 1, "Antares", 0.96},
			{3, 3, "Graffias", 2.56},
			{5, 4, "Dschubba", 2.29},
			{6, 5, "Wei", 2.62},
			{8, 6, "Shaula", 1.62},
		},
		Connections: [][2]int{{0, 1}, {1, 2}, {2, 3}, {3, 4}},
	},
	{
		Name:          "Lyra",
		Abbr:          "Lyr",
		Hemisphere:    "north",
		Season:        "summer",
		BrightestStar: "Vega (Alpha Lyrae, 0.03 mag)",
		Mythology:     "Lyra represents the lyre of Orpheus, the legendary musician and poet. After his death, the lyre was placed among the stars.",
		NotableObjs:   "Ring Nebula (M57), Vega (brightest star)",
		Stars: []Star{
			{5, 1, "Vega", 0.03},
			{3, 4, "Sheliak", 3.25},
			{7, 4, "Sulafat", 3.24},
		},
		Connections: [][2]int{{0, 1}, {0, 2}, {1, 2}},
	},
}

func getConstellation(query string) *Constellation {
	lower := strings.ToLower(query)
	for i := range constellations {
		if strings.ToLower(constellations[i].Name) == lower || strings.ToLower(constellations[i].Abbr) == lower {
			return &constellations[i]
		}
	}
	return nil
}

func listAll(hemisphere string) []Constellation {
	var result []Constellation
	for _, c := range constellations {
		if hemisphere == "all" || c.Hemisphere == hemisphere || c.Hemisphere == "both" {
			result = append(result, c)
		}
	}
	return result
}

func search(term string) []Constellation {
	var result []Constellation
	lower := strings.ToLower(term)
	for _, c := range constellations {
		if strings.Contains(strings.ToLower(c.Name), lower) ||
			strings.Contains(strings.ToLower(c.Abbr), lower) ||
			strings.Contains(strings.ToLower(c.Mythology), lower) {
			result = append(result, c)
		}
	}
	return result
}

func bySeason(season string) []Constellation {
	var result []Constellation
	for _, c := range constellations {
		if c.Season == season {
			result = append(result, c)
		}
	}
	return result
}

func drawLine(grid [][]string, x1, y1, x2, y2 int, char string) {
	dx := abs(x2 - x1)
	dy := abs(y2 - y1)
	sx := 1
	if x1 > x2 {
		sx = -1
	}
	sy := 1
	if y1 > y2 {
		sy = -1
	}
	err := dx - dy
	x, y := x1, y1
	for {
		if y >= 0 && y < len(grid) && x >= 0 && x < len(grid[0]) {
			if grid[y][x] != "•" {
				grid[y][x] = char
			}
		}
		if x == x2 && y == y2 {
			break
		}
		e2 := 2 * err
		if e2 > -dy {
			err -= dy
			x += sx
		}
		if e2 < dx {
			err += dx
			y += sy
		}
	}
}

func drawMap(query string) string {
	c := getConstellation(query)
	if c == nil {
		return fmt.Sprintf("❌ Constellation '%s' not found.", query)
	}
	maxX, maxY := 0, 0
	for _, s := range c.Stars {
		if s.X > maxX {
			maxX = s.X
		}
		if s.Y > maxY {
			maxY = s.Y
		}
	}
	maxX += 2
	maxY += 2
	grid := make([][]string, maxY)
	for i := range grid {
		grid[i] = make([]string, maxX)
		for j := range grid[i] {
			grid[i][j] = " "
		}
	}
	for _, s := range c.Stars {
		grid[s.Y][s.X] = "•"
	}
	for _, conn := range c.Connections {
		s1 := c.Stars[conn[0]]
		s2 := c.Stars[conn[1]]
		drawLine(grid, s1.X, s1.Y, s2.X, s2.Y, "·")
	}
	result := fmt.Sprintf("\n🌟 Constellation: %s (%s)\n\n", c.Name, c.Abbr)
	for _, row := range grid {
		result += "  " + strings.Join(row, "") + "\n"
	}
	result += "\n"
	for _, s := range c.Stars {
		result += fmt.Sprintf("  %s (mag %.2f)\n", s.Name, s.Mag)
	}
	return result
}

func info(query string) string {
	c := getConstellation(query)
	if c == nil {
		return fmt.Sprintf("❌ Constellation '%s' not found.", query)
	}
	return fmt.Sprintf("\n🌟 Constellation: %s\nAbbreviation: %s\nHemisphere: %s\nVisible Season: %s\nBrightest Star: %s\nMythology: %s\nNotable Objects: %s\n",
		c.Name, c.Abbr, c.Hemisphere, strings.Title(c.Season), c.BrightestStar, c.Mythology, c.NotableObjs)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: star_map <command> [options]")
		return
	}
	cmd := os.Args[1]

	switch cmd {
	case "list":
		listCmd := flag.NewFlagSet("list", flag.ExitOnError)
		hemisphere := listCmd.String("hemisphere", "all", "north, south, or all")
		listCmd.Parse(os.Args[2:])
		results := listAll(*hemisphere)
		fmt.Printf("\n📋 Constellations (%d):\n", len(results))
		for _, c := range results {
			fmt.Printf("  %s (%s) – %s\n", c.Name, c.Abbr, c.Hemisphere)
		}

	case "map":
		if len(os.Args) < 3 {
			fmt.Println("Error: need a constellation name or abbreviation")
			return
		}
		fmt.Print(drawMap(os.Args[2]))

	case "info":
		if len(os.Args) < 3 {
			fmt.Println("Error: need a constellation name or abbreviation")
			return
		}
		fmt.Print(info(os.Args[2]))

	case "season":
		if len(os.Args) < 3 {
			fmt.Println("Error: need a season (spring, summer, autumn, winter)")
			return
		}
		season := os.Args[2]
		results := bySeason(season)
		fmt.Printf("\n🌿 Constellations visible in %s:\n", strings.Title(season))
		for _, c := range results {
			fmt.Printf("  %s (%s)\n", c.Name, c.Abbr)
		}

	case "search":
		if len(os.Args) < 3 {
			fmt.Println("Error: need a search term")
			return
		}
		term := os.Args[2]
		results := search(term)
		if len(results) > 0 {
			fmt.Printf("\n🔍 Found %d constellation(s):\n", len(results))
			for _, c := range results {
				fmt.Printf("  %s (%s) – %s...\n", c.Name, c.Abbr, c.Mythology[:min(60, len(c.Mythology))])
			}
		} else {
			fmt.Println("❌ No matches found.")
		}

	default:
		fmt.Println("Unknown command. Use list, map, info, season, search.")
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: StarMap.cs
// StarMap.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Star
{
    public int X { get; set; }
    public int Y { get; set; }
    public string Name { get; set; }
    public double Mag { get; set; }
}

class Constellation
{
    public string Name { get; set; }
    public string Abbr { get; set; }
    public string Hemisphere { get; set; }
    public string Season { get; set; }
    public string BrightestStar { get; set; }
    public string Mythology { get; set; }
    public string NotableObjects { get; set; }
    public List<Star> Stars { get; set; } = new List<Star>();
    public List<int[]> Connections { get; set; } = new List<int[]>();
}

class StarMap
{
    private static readonly List<Constellation> CONSTELLATIONS = new List<Constellation>
    {
        new Constellation
        {
            Name = "Orion", Abbr = "Ori", Hemisphere = "both", Season = "winter",
            BrightestStar = "Betelgeuse (Alpha Orionis, 0.42 mag)",
            Mythology = "Orion, a mighty hunter in Greek mythology, was placed among the stars by Zeus. He is depicted as a hunter with his belt, sword, and club.",
            NotableObjects = "Orion Nebula (M42), Horsehead Nebula, Barnard's Loop",
            Stars = new List<Star>
            {
                new Star{X=4,Y=1,Name="Betelgeuse",Mag=0.42},
                new Star{X=5,Y=4,Name="Bellatrix",Mag=1.64},
                new Star{X=3,Y=5,Name="Mintaka",Mag=2.23},
                new Star{X=5,Y=5,Name="Alnilam",Mag=1.69},
                new Star{X=7,Y=5,Name="Alnitak",Mag=1.88},
                new Star{X=2,Y=8,Name="Saiph",Mag=2.06},
                new Star{X=6,Y=9,Name="Rigel",Mag=0.12}
            },
            Connections = new List<int[]> { new[]{0,1}, new[]{0,2}, new[]{0,3}, new[]{0,4}, new[]{0,5}, new[]{0,6},
                new[]{1,2}, new[]{2,3}, new[]{3,4}, new[]{4,5}, new[]{5,6}, new[]{2,4} }
        },
        new Constellation
        {
            Name = "Ursa Major", Abbr = "UMa", Hemisphere = "north", Season = "spring",
            BrightestStar = "Alioth (Epsilon Ursae Majoris, 1.77 mag)",
            Mythology = "Ursa Major represents the Great Bear. In Greek mythology, Callisto was transformed into a bear by Hera and placed in the sky by Zeus.",
            NotableObjects = "M81, M82 (Bode's Galaxy and Cigar Galaxy), Owl Nebula (M97)",
            Stars = new List<Star>
            {
                new Star{X=1,Y=1,Name="Dubhe",Mag=1.79},
                new Star{X=3,Y=2,Name="Merak",Mag=2.34},
                new Star{X=2,Y=4,Name="Phecda",Mag=2.41},
                new Star{X=4,Y=4,Name="Megrez",Mag=3.31},
                new Star{X=5,Y=5,Name="Alioth",Mag=1.77},
                new Star{X=7,Y=4,Name="Mizar",Mag=2.04},
                new Star{X=9,Y=3,Name="Alkaid",Mag=1.86}
            },
            Connections = new List<int[]> { new[]{0,1}, new[]{1,2}, new[]{2,3}, new[]{3,4}, new[]{4,5}, new[]{5,6} }
        },
        new Constellation
        {
            Name = "Cassiopeia", Abbr = "Cas", Hemisphere = "north", Season = "autumn",
            BrightestStar = "Schedar (Alpha Cassiopeiae, 2.24 mag)",
            Mythology = "Cassiopeia was the vain queen of Ethiopia who boasted about her beauty, leading to the sacrifice of her daughter Andromeda. She was placed in the sky as punishment.",
            NotableObjects = "Cassiopeia A (supernova remnant), NGC 457 (open cluster)",
            Stars = new List<Star>
            {
                new Star{X=2,Y=1,Name="Segin",Mag=3.37},
                new Star{X=3,Y=3,Name="Ruchbah",Mag=2.68},
                new Star{X=5,Y=2,Name="Schedar",Mag=2.24},
                new Star{X=6,Y=4,Name="Navi",Mag=2.47},
                new Star{X=8,Y=3,Name="Caph",Mag=2.28}
            },
            Connections = new List<int[]> { new[]{0,1}, new[]{1,2}, new[]{2,3}, new[]{3,4}, new[]{0,2}, new[]{2,4} }
        },
        new Constellation
        {
            Name = "Scorpius", Abbr = "Sco", Hemisphere = "south", Season = "summer",
            BrightestStar = "Antares (Alpha Scorpii, 0.96 mag)",
            Mythology = "Scorpius represents the scorpion that killed Orion. According to mythology, the scorpion was placed on the opposite side of the sky from Orion.",
            NotableObjects = "Antares (red supergiant), Ptolemy's Cluster (M7), Butterfly Cluster (M6)",
            Stars = new List<Star>
            {
                new Star{X=2,Y=1,Name="Antares",Mag=0.96},
                new Star{X=3,Y=3,Name="Graffias",Mag=2.56},
                new Star{X=5,Y=4,Name="Dschubba",Mag=2.29},
                new Star{X=6,Y=5,Name="Wei",Mag=2.62},
                new Star{X=8,Y=6,Name="Shaula",Mag=1.62}
            },
            Connections = new List<int[]> { new[]{0,1}, new[]{1,2}, new[]{2,3}, new[]{3,4} }
        },
        new Constellation
        {
            Name = "Lyra", Abbr = "Lyr", Hemisphere = "north", Season = "summer",
            BrightestStar = "Vega (Alpha Lyrae, 0.03 mag)",
            Mythology = "Lyra represents the lyre of Orpheus, the legendary musician and poet. After his death, the lyre was placed among the stars.",
            NotableObjects = "Ring Nebula (M57), Vega (brightest star)",
            Stars = new List<Star>
            {
                new Star{X=5,Y=1,Name="Vega",Mag=0.03},
                new Star{X=3,Y=4,Name="Sheliak",Mag=3.25},
                new Star{X=7,Y=4,Name="Sulafat",Mag=3.24}
            },
            Connections = new List<int[]> { new[]{0,1}, new[]{0,2}, new[]{1,2} }
        }
    };

    static Constellation GetConstellation(string query)
    {
        var lower = query.ToLower();
        return CONSTELLATIONS.FirstOrDefault(c => c.Name.ToLower() == lower || c.Abbr.ToLower() == lower);
    }

    static List<Constellation> ListAll(string hemisphere)
    {
        return CONSTELLATIONS.Where(c => hemisphere == "all" || c.Hemisphere == hemisphere || c.Hemisphere == "both").ToList();
    }

    static List<Constellation> Search(string term)
    {
        var lower = term.ToLower();
        return CONSTELLATIONS.Where(c =>
            c.Name.ToLower().Contains(lower) ||
            c.Abbr.ToLower().Contains(lower) ||
            c.Mythology.ToLower().Contains(lower)
        ).ToList();
    }

    static List<Constellation> BySeason(string season)
    {
        return CONSTELLATIONS.Where(c => c.Season == season).ToList();
    }

    static void DrawLine(char[][] grid, int x1, int y1, int x2, int y2, char ch)
    {
        int dx = Math.Abs(x2 - x1);
        int dy = Math.Abs(y2 - y1);
        int sx = x1 < x2 ? 1 : -1;
        int sy = y1 < y2 ? 1 : -1;
        int err = dx - dy;
        int x = x1, y = y1;
        while (true)
        {
            if (y >= 0 && y < grid.Length && x >= 0 && x < grid[0].Length)
                if (grid[y][x] != '•') grid[y][x] = ch;
            if (x == x2 && y == y2) break;
            int e2 = 2 * err;
            if (e2 > -dy) { err -= dy; x += sx; }
            if (e2 < dx) { err += dx; y += sy; }
        }
    }

    static string DrawMap(string query)
    {
        var c = GetConstellation(query);
        if (c == null) return $"❌ Constellation '{query}' not found.";
        int maxX = 0, maxY = 0;
        foreach (var s in c.Stars)
        {
            if (s.X > maxX) maxX = s.X;
            if (s.Y > maxY) maxY = s.Y;
        }
        maxX += 2; maxY += 2;
        var grid = new char[maxY][];
        for (int i = 0; i < maxY; i++) grid[i] = new string(' ', maxX).ToCharArray();
        foreach (var s in c.Stars) grid[s.Y][s.X] = '•';
        foreach (var conn in c.Connections)
        {
            var s1 = c.Stars[conn[0]];
            var s2 = c.Stars[conn[1]];
            DrawLine(grid, s1.X, s1.Y, s2.X, s2.Y, '·');
        }
        var result = $"\n🌟 Constellation: {c.Name} ({c.Abbr})\n\n";
        foreach (var row in grid) result += "  " + new string(row) + "\n";
        result += "\n";
        foreach (var s in c.Stars) result += $"  {s.Name} (mag {s.Mag:F2})\n";
        return result;
    }

    static string Info(string query)
    {
        var c = GetConstellation(query);
        if (c == null) return $"❌ Constellation '{query}' not found.";
        return $"\n🌟 Constellation: {c.Name}\nAbbreviation: {c.Abbr}\nHemisphere: {c.Hemisphere}\nVisible Season: {char.ToUpper(c.Season[0]) + c.Season.Substring(1)}\nBrightest Star: {c.BrightestStar}\nMythology: {c.Mythology}\nNotable Objects: {c.NotableObjects}";
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: StarMap <command> [options]");
            return;
        }
        var parsed = ParseArgs(args);
        string cmd = args[0];
        switch (cmd)
        {
            case "list":
                string hemisphere = parsed.GetValueOrDefault("hemisphere", "all");
                var results = ListAll(hemisphere);
                Console.WriteLine($"\n📋 Constellations ({results.Count}):");
                foreach (var c in results) Console.WriteLine($"  {c.Name} ({c.Abbr}) – {c.Hemisphere}");
                break;
            case "map":
                if (args.Length < 2) { Console.WriteLine("Error: need a constellation name or abbreviation"); return; }
                Console.Write(DrawMap(args[1]));
                break;
            case "info":
                if (args.Length < 2) { Console.WriteLine("Error: need a constellation name or abbreviation"); return; }
                Console.Write(Info(args[1]));
                break;
            case "season":
                if (args.Length < 2) { Console.WriteLine("Error: need a season (spring, summer, autumn, winter)"); return; }
                var seasonResults = BySeason(args[1]);
                Console.WriteLine($"\n🌿 Constellations visible in {char.ToUpper(args[1][0]) + args[1].Substring(1)}:");
                foreach (var c in seasonResults) Console.WriteLine($"  {c.Name} ({c.Abbr})");
                break;
            case "search":
                if (args.Length < 2) { Console.WriteLine("Error: need a search term"); return; }
                var searchResults = Search(args[1]);
                if (searchResults.Any())
                {
                    Console.WriteLine($"\n🔍 Found {searchResults.Count} constellation(s):");
                    foreach (var c in searchResults)
                        Console.WriteLine($"  {c.Name} ({c.Abbr}) – {(c.Mythology.Length > 60 ? c.Mythology.Substring(0,60) + "..." : c.Mythology)}");
                }
                else Console.WriteLine("❌ No matches found.");
                break;
            default:
                Console.WriteLine("Unknown command. Use list, map, info, season, search.");
                break;
        }
    }

    static Dictionary<string, string> ParseArgs(string[] args)
    {
        var dict = new Dictionary<string, string>();
        for (int i = 1; i < args.Length; i++)
        {
            if (args[i].StartsWith("--") && i + 1 < args.Length)
                dict[args[i].Substring(2)] = args[++i];
        }
        return dict;
    }
}
